### PR TITLE
Freeze dictionary keys to make them hashable

### PR DESCRIPTION
**Pull request by nielstron**
Alternative solution for #227

### CODE REVIEW

**Comment by nielstron** [comment]
This should fix https://github.com/OpShin/opshin-pioneer-program/pull/43

**Comment by nielstron** [comment]
Only codecov failed uploading, possibly due to the CI being incorrectly set up

**Comment by juliusfrost** [comment]
It looks like there's double recursion in _freeze and _helper, which squares the search depth. Maybe merge the two to freeze just the outer data type, since we can assume the inner values are already frozen.

**Review by cffls — APPROVED** [review]
This is awesome! 🚀

**Comment by nielstron** [comment]
> It looks like there's double recursion in _freeze and _helper, which squares the search depth. Maybe merge the two to freeze just the outer data type, since we can assume the inner values are already frozen.

You're right, the Seth elements and dictionary keys do not need to be frozen again, they can be assumed to be frozen.

A frozenhelper function also sounds interesting 🤔

**Comment by juliusfrost** [comment]
I'm looking at this again from my computer (as opposed to my phone), I think I meant to say merge `_dfs` with `_frozen`

**Comment by juliusfrost** [comment]
@nielstron I made a quick edit here: https://github.com/OpShin/pycardano/pull/3

**Comment by juliusfrost** [comment]
This might be a good opportunity to use hypothesis to generate arbitrary CBORSerializable types

**Comment by juliusfrost** [comment]
If we don't want to freeze all the time we could set a freeze argument and only freeze when true, then set it true recursively for keys but default false otherwise

**Comment by juliusfrost** [comment]
> If we don't want to freeze all the time we could set a freeze argument and only freeze when true, then set it true recursively for keys but default false otherwise

I implemented this here: https://github.com/OpShin/pycardano/pull/4

File: pycardano/serialization.py [hunk 1]
```diff
@@ -214,46 +230,42 @@ def to_primitive(self) -> Primitive:
...
-                return new_result
+                    _dict[_dfs(k, freeze=True)] = _dfs(v, freeze)
+                if freeze:
+                    return frozendict(_dict)
+                return _dict
             elif isinstance(value, set):
-                return {_helper(v) for v in value}
-            elif isinstance(value, frozenset):
-                return frozenset({_helper(v) for v in value})
+                _set = set(_dfs(v, freeze) for v in value)
```
Comment (nielstron) [inline pycardano/serialization.py]:
```suggestion
                _set = set(_dfs(v, freeze=True) for v in value)
```

**Comment by juliusfrost** [comment]
https://github.com/OpShin/pycardano/pull/5
Fixes `TypeError: unhashable type: 'list'`

**Comment by juliusfrost** [comment]
This recent [PR](https://github.com/Marco-Sulla/python-frozendict/pull/70) in frozendict fixes the mypy issue when I tested it locally. Hope they make a new release soon

**Comment by juliusfrost** [comment]
https://github.com/OpShin/pycardano/pull/6 fixes issues with `poetry run mypy --install-types --non-interactive pycardano`

**Review by juliusfrost — APPROVED** [review]

**Comment by juliusfrost** [comment]
@cffls This should be ready to merge :)

**Comment by cffls** [comment]
Awesome, thanks! LGTM.

**Pull request merged**